src/Kestrel.Core/Internal/Http/HttpProtocol.cs

```diff
@@ -715,29 +717,25 @@ protected Task FireOnCompleted()
                 _onCompleted = null;
             }
 
-            if (onCompleted == null)
-            {
-                return Task.CompletedTask;
-            }
-            else
-            {
-                return FireOnCompletedAwaited(onCompleted);
-            }
+            FireOnCompletedAwaited(onCompleted);
         }
 
-        private async Task FireOnCompletedAwaited(Stack<KeyValuePair<Func<object, Task>, object>> onCompleted)
+        private void FireOnCompletedAwaited(Stack<KeyValuePair<Func<object, Task>, object>> onCompleted)
         {
-            foreach (var entry in onCompleted)
+            Task.Run(async () =>
```

Comment:
We don't want to fire and forget the OnCompleted callbacks. It's important to wait on the callbacks in the request processing loop to ensure that we don't dispose the HttpContext prior to the callbacks completing.

I think the fix for #2035 will be as simple as moving the call to `await FireOnCompleted();` to after the call to `await ProduceEnd();`, but I haven't tried it yet.